src/util/auth/hs_auth.ts

```diff
@@ -94,58 +86,42 @@ export class RequestAuthentication {
 		});
 	}
 
-	/**
-   * checkLoggedIn middleware calls the authentication function to
-   * validate the users request, is promise rejected, then redirect to login
-   * @param req Request object from express
-   * @param res Response object from express
-   * @param next Next Callback function used to move to the next middleware
-   */
-	public async checkLoggedIn(req: Request, res: Response, next: NextFunction): Promise<void> {
-		let user: User;
-		try {
-			user = await this.authenticate(req, res);
-		} catch (err) {
-			// Either user was not authenticated, or an error occured during authentication
-			// In both cases we redirect them to the login
-			const queryParam: string = querystring.encode({
-				returnto: `${getConfig().hs.applicationUrl}${req.originalUrl}`
-			});
-			res.redirect(`${getConfig().hs.authUrl}/login?${queryParam}`);
-			return;
-		}
-		res.locals.authLevel = user.authLevel;
-		return next();
-	}
+	public withAuthMiddleware(router: RouterInterface, operationHandler: ExpressOpHandlerFunction): AuthMiddlewareFunction<unknown> {
+		return async (req: Request, res: Response, next: NextFunction): Promise<unknown> => {
+			const userAuth = this.authenticate(req, res);
 
-	public checkAuthLevel(req: Request, res: Response, user: User|undefined, requiredAuth: AuthLevel): boolean {
-		if (!user || user.authLevel < requiredAuth) {
-			const queryParam: string = querystring.encode({ returnto: `${getConfig().hs.applicationUrl}${req.originalUrl}` });
-			res.redirect(`${getConfig().hs.authUrl}/login?${queryParam}`);
-			return false;
-		}
-		return true;
+			const requestedUri = this.getUriFromRequest(router, operationHandler, req);
+			const resourceAuth = this.authApi.getAuthorizedResources(this.getAuthToken(req), [requestedUri]);
+
+			try {
+				const [user, permissions] = await Promise.all([userAuth, resourceAuth]);
+				if (permissions.length === 0) {
+					return this.handleUnauthorized(req, res);
+				}
+				req.user = user;
+			} catch (err) {
+				return this.handleUnauthorized(req, res);
+			}
+
+			return operationHandler(req, res, next);
+		};
 	}
 
-	public checkIsAttendee(req: Request, res: Response, next: NextFunction): void {
-		if (this.checkAuthLevel(req, res, req.user as User, AuthLevel.Attendee)) {
-			res.locals.isAttendee = true;
-			return next();
-		}
+	public handleUnauthorized(req: Request, res: Response): void {
+		const queryParam: string = querystring.encode({ returnto: `${getConfig().hs.applicationUrl}${req.originalUrl}` });
+		res.redirect(`${getConfig().hs.authUrl}/login?${queryParam}`);
 	}
 
-	public checkIsVolunteer(req: Request, res: Response, next: NextFunction): void {
-		if (this.checkAuthLevel(req, res, req.user as User, AuthLevel.Volunteer)) {
-			res.locals.isVolunteer = true;
-			return next();
-		}
+	public getAuthToken(req: Request): string {
```

Comment:
Could you rename this to `getUserAuthToken`? We're going to use service tokens as well, so it might lead to some confusion later.

Reply:
Good point!